java/ql/src/semmle/code/java/security/XPath.qll

```diff
@@ -0,0 +1,56 @@
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.security.XmlParsers
+
+/**
+ * An abstract type representing a call to interpret XPath expressions.
+ */
+class XPathSink extends MethodAccess {
+  abstract Expr getSink();
+}
+
+/** The class `javax.xml.xpath.XPath` */
+class XPath extends RefType {
+  XPath() { this.hasQualifiedName("javax.xml.xpath", "XPath") }
+}
+
+/** A call to `XPath.Evaluate` or `XPath.compile` */
+class XPathEvaluateOrCompile extends XPathSink {
+  XPathEvaluateOrCompile() {
+    exists(Method m | this.getMethod() = m and m.getDeclaringType() instanceof XPath |
+      m.hasName("evaluate")
+      or
+      m.hasName("compile")
+    )
+  }
+
+  override Expr getSink() { result = this.getArgument(0) }
+}
+
+/** The class `org.dom4j.Node` */
+class Dom4JNode extends RefType {
+  Dom4JNode() { this.hasQualifiedName("org.dom4j", "Node") }
+}
+
+/** A call to `Node.selectNodes` or `Node.selectSingleNode` */
+class NodeSelectNodes extends XPathSink {
+  NodeSelectNodes() {
+    exists(Method m | this.getMethod() = m and m.getDeclaringType() instanceof Dom4JNode |
+      m.hasName("selectNodes") or m.hasName("selectSingleNode")
```

Comment:
This can be simplified using set literals.

Reply:
Good suggestion, thanks! Added it in d592b1c